### DISCUSSION TITLE
How to use gtk::Adjustment?

### DISCUSSION THREAD

**Discussion by eminfedar** [Relm4/Relm4#503] [Q&A] [answered]
I want to use [gtk::Adjustment](https://gtk-rs.org/gtk4-rs/stable/latest/docs/gtk4/struct.Adjustment.html) in [gtk::SpinButton](https://gtk-rs.org/gtk4-rs/stable/latest/docs/gtk4/struct.SpinButton.html):
```rust
use gtk::prelude::AdjustmentExt;

view! {
// ...

gtk::SpinButton {
    set_adjustment: &gtk::Adjustment {
        lower: 1,
        upper: 120,
        value: 15,
        step_increment: 1,
        page_increment: 10,
    },
},

}
```

But I got this error message:
```rust
error[E0560]: struct `Adjustment` has no field named `lower`
  --> src/ui2.rs:89:25
   |
89 |                         lower: 1,
   |                         ^^^^^ `Adjustment` does not have this field

error[E0560]: struct `Adjustment` has no field named `upper`
  --> src/ui2.rs:90:25
   |
90 |                         upper: 120,
   |                         ^^^^^ `Adjustment` does not have this field

...
```

What is the correct way to use gtk::Adjustment in view! macro?

**Comment by AaronErhardt** [comment, marked as answer]
`gtk::Adjustment` is a type of the GTK library, which means that it - unlike regular Rust structs - doesn't have fields, but you have to set properties (which is a concept of GObject) and properties can only be set by calling methods.

This means, there are several ways to solve this problem.

1. Use the builder type. The gtk4 crate provides convenient builders for initializing widgets and other types. 

```rust
gtk::SpinButton {
    set_adjustment: &gtk::Adjustment::builder().
        .lower(1)
        .upper(120)
        .value(15)
        .step_increment(1)
        .page_increment(10)
        .build() // This converts the builder into gtk::Adjustment again
    },
}
```

2. Crate a new variable in the macro. The macro has two options for setting values. The colon `:` is used for assigning values similar to struct initialization in regular Rust code. The `= Type { ... }` syntax on the other side creates a new variable on which we can call methods. It uses the `Default` implementation of the type under the hood btw. unless you specify a method like `= Type::new() { ... }`.

Inside of the brackets `{}` you can then set the properties using the `AdjustmentExt` trait.

```rust
gtk::SpinButton {
    set_adjustment = &gtk::Adjustment {
        set_lower: 1,
        set_upper: 120,
        set_value: 15,
        set_step_increment: 1,
        set_page_increment: 10,
    },
}
```

The generated code from the macro will look roughly like this btw.

```rust
let spinner = gtk::SpinButton::default();
let adjustment = gtk::Adjustment::default();

spinner.set_adjustment(&adjustment);

adjustment.set_lower(1);
adjustment.set_upper(120);
adjustment.set_value(5);
// ...
```
